music21/metadata/__init__.py

```diff
@@ -142,115 +238,376 @@ class Metadata(base.Music21Object):
-        for abbreviation, workId in self.workIdAbbreviationDict.items():
-            # abbreviation = workIdToAbbreviation(id)
-            if workId in keywords:
-                self._workIds[workId] = Text(keywords[workId])
-            elif abbreviation in keywords:
-                self._workIds[workId] = Text(keywords[abbreviation])
-            else:
-                self._workIds[workId] = None
+    def setCustom(self, key: str, value: t.Union[t.Any, t.List[t.Any]]):
+        self._set(key, value, isCustom=True)
+
+    def getAllNamedValues(self, skipContributors=False) -> t.List[t.Tuple[str, t.Any]]:
+        '''
+        Returns all values stored in this metadata as a list of (nsKey, value) tuples.
+        nsKeys with multiple values will appear multiple times in the list (rather
+        than appearing once, with a value that is a list of values).
+        The tuple's first element is either of the form 'namespace:name', or a
+        custom key (with no form at all).
+
+        >>> md = metadata.Metadata()
+        >>> md.add('composer', 'Jeff Bowen')
+        >>> md.add('librettist', 'Hunter Bell')
+        >>> md.add('title', '[title of show]')
+        >>> md.addCustom('excerpt-start-measure', 1234)
+        >>> all = md.getAllNamedValues()
+        >>> all
+        [('marcrel:CMP', <music21.metadata.primitives.Contributor composer:Jeff Bowen>),
+        ('marcrel:LBT', <music21.metadata.primitives.Contributor librettist:Hunter Bell>),
+        ('dcterms:title', <music21.metadata.primitives.Text [title of show]>),
+        ('excerpt-start-measure', <music21.metadata.primitives.Text 1234>)]
+        >>> all = md.getAllNamedValues(skipContributors=True)
+        >>> all
+        [('dcterms:title', <music21.metadata.primitives.Text [title of show]>),
+        ('excerpt-start-measure', <music21.metadata.primitives.Text 1234>)]
+        '''
+        allOut: t.List[t.Tuple[str, t.Any]] = []
+
+        for nsKey, value in self._metadata.items():
+            if skipContributors and self._isContributorNSKey(nsKey):
+                continue
+
+            for v in value:
+                allOut.append((nsKey, v))
+
+        return allOut
+
+    def getAllContributorNamedValues(self) -> t.List[t.Tuple[str, t.Any]]:
+        '''
+        Returns all contributors stored in this metadata as a list of (nsKey, value) tuples.
+        The tuple's first element will be of the form 'namespace:name'.
+
+        >>> md = metadata.Metadata()
+        >>> md.addCustom('composer', 'Not a contributor')
+        >>> md.add('title', '[title of show]')
+        >>> md.addCustom('excerpt-start-measure', 1234)
+        >>> md.add('composer', ['Ludwig von Beethoven', 'Wolfgang Amadeus Mozart'])
+        >>> md.add('librettist', 'Marie Červinková-Riegrová')
+        >>> all = md.getAllContributorNamedValues()
+        >>> len(all)
+        3
+        >>> all[0]
+        ('marcrel:CMP', <music21.metadata.primitives.Contributor composer:Ludwig von Beethoven>)
```

Comment:
where is `marcrel` cross-referenced?  I had heard of MARC but never marcrel before this review.  'Composer' is a standard value in many, many Metadata formats, why is 'marcrel' getting pride of place here?

Reply:
is marcrel = Dublin Core?

Reply:
It refers to the MARC21 Relator Codes: 

https://www.loc.gov/marc/relators/relaterm.html

They are available for use in DC: https://www.dublincore.org/specifications/dublin-core/relators/

The `marcrel:` prefix should expand to a value of `http://www.loc.gov/loc.terms/relators/`. This would make the full URI e.g., http://www.loc.gov/loc.terms/relators/CMP for this relationship type.

See also: https://memory.loc.gov/diglib/loc.terms/relators/dc-contributor.html

Reply:
Some background for the decisions I made about Dublin Core and namespaces:

The original Dublin Core "elements" namespace (usually abbreviated "dc:") has only a very small number of things (contributor, coverage, creator, date, description, format, identifier, language, publisher, relation, rights, source, subject, title, type).  

The Dublin Core "terms" namespace (usually abbreviated "dcterms:", or sometimes "dcterm:") showed up a few years later, and adds a bunch more good metadata, but still very few different contributors.  

The MARC Relator codes are an extension of Dublin Core (usually abbreviated "marcrel:") that specifically adds a full list of contributors.   Currently the version of the Marc Relator codes I am using has a lot of contributors that have nothing to do with music (e.g. "Interviewer"), but I chose to err on the side of inclusion (we can discuss exactly where the inclusion line should be).  There are longer MARC Relator code lists I have seen, though, so maybe there are other codes I should add during that deletion exercise...  In any case, these are all officially available for use in Dublin Core (dcterms + marcrel) and then I've also added a bunch of "humdrum:" codes (these are not officially Dublin Core extensions, but they provide the rest of the workIds/contributors supported by music21 v7).  And then I added one more recently: "subtitle", which I gave the "mei:" prefix, since that's the parser that wanted it (and no-one else had it).

Reply:
Further info on the subset of MARC Relator codes I chose: I picked all the ones that were refinements of 'dcterms:contributor'.  There are a bunch more I didn't use that aren't contributors at all, just people who were associated in some way.  e.g. lender, libelee, etc.

Reply:
I'll think more -- with some other search terms, there seem to be more support for MARC Relator than I initially thought, AND it fits better with the Humdrum terms than I thought.  But it's still doesn't have the whole tutorial, python ecosystem, etc. that I would like.  Maybe we can support a subset of Relator that fits better with Humdrum/Music and remove terms like Animator, etc.  (Interviewer might be okay, for an ethnographic transcription w/ music).

Reply:
I'm going to mark this resolved, even though I know we will come back to this once the rest of the review is complete.